torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+
+
+def infer_mesh(slice_ranks, world_size):
+  '''Infers a rectangular mesh topology from a slice in the mesh.
+
+    Example, given world size 12 and a slice like the following:
+        [1, 5, 9]
+    this function infers the following mesh:
+        [[0, 4, 8],
+         [1, 5, 9],
+         [2, 6, 10],
+         [3, 7, 11]]
+
+    We only support ractangular meshes.
+    '''
+  slice_len = len(slice_ranks)
+  if world_size % slice_len != 0:
+    raise ValueError(
+        'Given slice length doesn\'t equal to a side of a rectangle. '
+        f'World size: {world_size}, slice length: {slice_len}')
+
+  # ensure the ranks are a correct range.
+  start = slice_ranks[0]
+  step = slice_ranks[1] - slice_ranks[0]
+  stop = start + step * slice_len
+  expected_ranks = list(range(start, stop, step))
+  if slice_ranks != expected_ranks:
+    raise ValueError(
+        f'Given slice isn\'t a range with evenly distributed steps: {slice_ranks}'
+    )
+
+  num_slices = int(world_size / slice_len)
+  world = list(range(world_size))
+  mesh = []
+  if step == 1:
+    # Horizontal case.
+    if start % slice_len != 0 or start >= world_size:
+      raise ValueError('Given horizontal slice doesn\'t have a correct start: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    for i in range(num_slices):
+      slice_start = i * slice_len
+      slice_stop = (i + 1) * slice_len
+      slice_idx = slice(slice_start, slice_stop, step)
+      mesh.append(world[slice_idx])
+  else:
+    # Vertical case.
+    if start >= num_slices:
+      raise ValueError('Given vertical slice doesn\'t have a correct start: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    if step != num_slices:
+      raise ValueError('Given vertical slice doesn\'t have a correct step: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    for i in range(num_slices):
+      slice_start = i
+      slice_stop = i + slice_len * step
+      slice_idx = slice(slice_start, slice_stop, step)
+      mesh.append(world[slice_idx])
+  assert slice_ranks in mesh
+
+  return mesh
+
+
+def new_xla_process_group(ranks=None,
+                          timeout=dist.default_pg_timeout,
+                          backend=None):
+  pg = _orig_new_group_fn(ranks, timeout, backend)
+  if isinstance(pg, ProcessGroupXla) and ranks is not None:
+    world_pg = dist.group.WORLD
+    if not isinstance(world_pg, ProcessGroupXla):
+      raise RuntimeError('xla backend requires the default ProcessGroup to be '
+                         'a ProcessGroupXla')
+
+    if isinstance(ranks, range):
+      ranks = list(ranks)
+
+    if ranks == list(range(world_pg.size())):
+      pg._mesh = [ranks]
+    elif len(ranks) == 1:
+      if ranks[0] not in range(world_pg.size()):
```

Comment:
Curious, what's the difference between `ranks=[1, 2, 3, 4]`, `ranks=[[1], [2], [3], [4]]`, and `ranks=[[1, 2], [3, 4]]`? Will there be any differences in interconnect bandwidth?

Reply:
Also, what's the contract here? Do we assume that each process will provide a unique rank id? I.e., process 0 sets `ranks=[[0]]`, process 0 sets `ranks=[[1]]`, and so on. If so, we do need to verify that?

Reply:
Here `ranks` is the input and it's a list of ranks like [1,2,3,4]. The `_mesh` is the output and is a 2-d array of ranks and in a regular rectangle shape (not zigzagged), e.g. [[1, 2], [3, 4]].

A mesh of `[[1, 2], [3, 4]]` means the CC ops will happen only within each group: 1 and 2 as one group and 3 and 4 as another group. No CC will happen across the groups.

Reply:
> A mesh of [[1, 2], [3, 4]] means the CC ops will happen only within each group: 1 and 2 as one group and 3 and 4 as another group. No CC will happen across the groups.

I see. This seems like a new behavior, i.e., creating multiple sub-groups in one `new_group` call. For other distributed backends, all ranks needs to run the following code to do that. 

```python
new_group(ranks=[1, 2], ...)  # rank 3 and 4 also need to run this even though they are not in the group
new_group(ranks=[3, 4], ...)  # rank 1 and 2 also need to run this even though they are not in the group
```

TBH, I think this new behavior is a great addition to the current `new_group` API. We can potentially add that to c10d as well (when this is done), so that you don't need to override `new_group`.